[msl-out] Make varyings' struct members unique

evahop opened the pull request:
Fixes #1490

Naming conflicts can occur when multiple structs get flattened into one varyings struct.
This moves the `namer.call` invocation into the flattening routine so that each member gets a unique name.

teoxoy reviewed (changes requested):
Thanks for the PR! It looks good but could you add a test for this? The one in the issue would work.

evahop commented:
Yes, I was going to add that there's also a `namer.namespace` method specifically for structs, It may be more appropriate, but I am not familiar enough with result types. I will try to use that instead, but no promises.

evahop commented:
Added tests, including one that confirms #1464 has also been fixed (not by me).
Abandoned  `namer.namespace` since `built-ins` don't necessarily end up in the struct, 
I just added a new `namer` solely for the locations.

evahop commented:
Added snapshot, including one that confirms #1464 has also been fixed (not by me).
Abandoned  `namer.namespace` and just ended up just added a new `namer` solely for the locations. It causes a bit of duplication, since I also stick them in `flattened_member_names` if you want.

evahop commented:
Added snapshot, which also confirms #1464 has been fixed (not by me).
Abandoned  `namer.namespace` and ended up just added a new `namer` solely for the locations. It causes a bit of duplication, since I also stick them in `flattened_member_names` if you want.

evahop commented:
Added the snapshot, which also confirms #1464 has been fixed (not by me).
Abandoned  `namer.namespace` and ended up just adding a new `namer` solely for the locations.

teoxoy commented on src/back/msl/writer.rs:
```diff
@@ -3704,6 +3713,16 @@ impl<W: Write> Writer<W> {
                                 member.ty,
                                 member.binding.as_ref(),
                             ));
+                            let name_key = &NameKey::StructMember(arg.ty, member_index);
+                            flattened_member_names.insert(
+                                NameKey::StructMember(arg.ty, member_index),
+                                match member.binding {
+                                    Some(crate::Binding::Location { .. }) => {
+                                        varyings_namer.call(&self.names[name_key])
+                                    }
+                                    _ => self.namer.call(&self.names[name_key]),
+                                },
+                            );
```
```suggestion
                            let name_key = NameKey::StructMember(arg.ty, member_index);
                            let name = match member.binding {
                                Some(crate::Binding::Location { .. }) => {
                                    varyings_namer.call(&self.names[&name_key])
                                }
                                _ => self.namer.call(&self.names[&name_key]),
                            };
                            flattened_member_names.insert(name_key, name);
```

teoxoy commented:
Seems like https://github.com/gfx-rs/naga/issues/1464 was fixed by https://github.com/gfx-rs/naga/pull/1766.

teoxoy reviewed (approved):
Good stuff!
